arch-riscv: This commit fixes bug in vfmv.f.s impl. in riscv

debjyoti0891 opened the pull request:
The existing implementation of vfmv instruction did not type cast the first element of the source vector, which caused the "freg" to interpret the result as a NaN.

With the type cast to f32, the value is correctly recognized as float and sign extended to be stored in the fd register.

Git issue: https://github.com/gem5/gem5/issues/827

Change-Id: Ibe9873910827594c0ec11cb51ac0438428c3b54e

Commit:
arch-riscv: This commit fixes bug in vfmv.v.s impl. in riscv

The existing implementation of vfmv instruction did not type cast
the first element of the source vector, which caused the "freg"
to interpret the result as a NaN.

With the type cast to f32, the value is correctly recognized as
float and sign extended to be stored in the fd register.

Git issue: https://github.com/gem5/gem5/issues/827

Change-Id: Ibe9873910827594c0ec11cb51ac0438428c3b54e

hnpl reviewed (commented):
Hi, thanks for the fix.

I wonder if we should always cast the value to float32. My understanding is that the values in a vector register could be a set of float16 values or a set of float64 values too. Maybe we need to obtain the type from VTYPE.

~However, my understanding of vfmv.v.s is that it should be floating point type agnostic as we are merely copying bits from 1 floating point reg to a vector reg.~

According to https://github.com/riscv/riscv-v-spec/issues/339, we'll need to check for NaN if SEW != FLEN.

debjyoti0891 commented:
Hi @hnpl ,

Thanks for the feedback. Indeed, the cast should be based on the SEW. I am curious if I could use the [spike reference implementation](https://github.com/riscv-software-src/riscv-isa-sim/blob/master/riscv/insns/vfmv_s_f.h) as a baseline.

For NaN box, I am curious about the interpretation. 
- Do I check if the value is NaN in element 0 for each datatype (fp8, fp16, fp32,fp64)? 
- If this is true, the result stored in the floating point register is a canonical nan (7fc00000).

rogerchang23424 commented on src/arch/riscv/isa/decoder.isa:
```diff
@@ -2746,7 +2746,7 @@ decode QUADRANT default Unknown::unknown() {
                         // (vm=0) of vfmv.f.s are reserved
                         0x1: VectorNonSplitFormat::vfmv_f_s({{
-                            freg_t fd = freg(Vs2_vu[0]);
+                            freg_t fd = freg(f32(Vs2_vu[0]));
```
From `template/vector_aarith.isa`

```
def template VectorFloatNonSplitDecodeBlock {{

switch(machInst.vtype8.vsew) {
case 0b010: return new %(class_name)s<float32_t>(machInst);
case 0b011: return new %(class_name)s<float64_t>(machInst);
default: GEM5_UNREACHABLE;
}
```

I think we also need to check the type of `vu` before casting the `f32` and `f64`. We can assume the `FLEN = 64` as we don't have `FLEN` configuration.

ivanaamit commented:
@debjyoti0891, can you address Roger's comment (https://github.com/gem5/gem5/pull/863#discussion_r1488792013)? We would like to merge this PR. Thank you.

debjyoti0891 commented:
> @debjyoti0891, can you address Roger's comment ([#863 (comment)](https://github.com/gem5/gem5/pull/863#discussion_r1488792013))? We would like to merge this PR. Thank you.

surely, I plan to wrap this up in the next week.

ivanaamit commented:
@debjyoti0891, a friendly reminder to address pending comment. Thanks.

tommarin committed:
arch-riscv: Generalize type casting in vfmv.f.s

Cast the first element of the source vector to the correct type
depending on the element width.

Change-Id: I192f71d781cde49cd5a3bc42df4a0651773158dd

rogerchang23424 reviewed (approved):
Thanks for updating

debjyoti0891 committed:
Merge branch 'develop' into vfmv_fix

debjyoti0891 commented:
@ivanaamit , a friendly reminder to merge the PR
